controllers/account/api/v1/debt_webhook.go

```diff
@@ -0,0 +1,149 @@
+	kubeSystemNamespace  = "kube-system"
+	defaultUserNamespace = "user-system"
+)
+
+var logger = logf.Log.WithName("debt-resource")
+
+//func (i *Debt) SetupWebhookWithManager(mgr ctrl.Manager) error {
+//	m := &DebtMutator{Client: mgr.GetClient()}
+//	return ctrl.NewWebhookManagedBy(mgr).
+//		For(i).
+//		WithDefaulter(m).
+//		Complete()
+//}
+
+//+kubebuilder:webhook:path=/mutate-v1-sealos-cloud,mutating=true,failurePolicy=ignore,groups="*",resources=*,verbs=create;update,versions=v1,name=mpod.kb.io,admissionReviewVersions=v1,sideEffects=None
+// +kubebuilder:object:generate=false
+
+type DebtValidate struct {
+	Client client.Client
+}
+
+func getUserNamespace() string {
+	userNamespace := os.Getenv("USER_NAMESPACE")
+	if userNamespace == "" {
+		return defaultUserNamespace
+	}
+	return userNamespace
+}
+
+var kubeSystemGroup, userSaGroup string
+
+func init() {
+	kubeSystemGroup = fmt.Sprintf("%ss:%s", saPrefix, kubeSystemNamespace)
+	userSaGroup = fmt.Sprintf("%ss:%s", saPrefix, getUserNamespace())
+}
+func (d DebtValidate) Handle(ctx context.Context, req admission.Request) admission.Response {
+	logger.Info("checking user", "userInfo", req.UserInfo, "req.Namespace", req.Namespace, "req.Name", req.Name, "req.gvrk", getGVRK(req))
+	for _, g := range req.UserInfo.Groups {
+		switch g {
+		// if user is kubernetes-admin, pass it
+		case mastersGroup:
+			logger.Info("pass for kubernetes-admin")
+			return admission.ValidationResponse(true, "")
+		case kubeSystemGroup:
+			logger.Info("pass for kube-system")
+			return admission.ValidationResponse(true, "")
+		case userSaGroup:
+			logger.Info("check for user", "user", req.UserInfo.Username, "ns: ", req.Namespace)
+			if isWriteList(req) {
+				return admission.ValidationResponse(true, "")
+			}
+			return checkOption(ctx, logger, d.Client, req.Namespace)
+		default:
+			// continue to check other groups
+			continue
+		}
+	}
+
+	// check is user ns
+	logger.Info("pass ", "req.Namespace", req.Namespace)
+	return admission.ValidationResponse(true, "")
+}
+
+func getGVRK(req admission.Request) string {
+	if req.Kind.Group == "" {
+		return fmt.Sprintf("%s.%s/%s", req.Resource.Resource, req.Kind.Kind, req.Kind.Version)
+	}
+	return fmt.Sprintf("%s.%s.%s/%s", req.Resource.Resource, req.Kind.Kind, req.Kind.Group, req.Kind.Version)
+}
+
+func isWriteList(req admission.Request) bool {
+	whitelists := os.Getenv("WHITELIST")
+	if whitelists == "" {
+		return false
+	}
+	writelist := strings.Split(whitelists, ",")
+	for _, w := range writelist {
+		if getGVRK(req) == w {
+			logger.Info("pass for whitelists", "gck", req.Kind.String(), "name", req.Name, "namespace", req.Namespace, "userinfo", req.UserInfo)
+			return true
+		}
+	}
+
+	return false
+}
+
+func checkOption(ctx context.Context, logger logr.Logger, c client.Client, nsName string) admission.Response {
+	nsList := &corev1.NamespaceList{}
+	if err := c.List(ctx, nsList, client.MatchingFields{"name": nsName}); err != nil {
```

Comment:
Wonder change `list` to `get` will have better performance?

Reply:
if you want to use cache,you need to use list and client.MatchingFields to use cache.There is no get func that can use cache.